org.eclipse.draw2d/src/org/eclipse/draw2d/EventListenerList.java

```diff
@@ -52,50 +50,8 @@ public synchronized <T> void addListener(Class<T> c, Object listener) {
 	 * @return whether this list contains a listener of type <i>c</i>
 	 */
 	public synchronized <T> boolean containsListener(Class<T> c) {
-		if (array == null) {
-			return false;
-		}
-		for (int i = 0; i < array.length; i += 2) {
-			if (array[i] == c) {
-				return true;
-			}
-		}
-		return false;
-	}
-
-	static class TypeIterator<T> implements Iterator<T> {
-		private final Object[] items;
-		private final Class<T> type;
-		private int index;
-
-		TypeIterator(Object[] items, Class<T> type) {
-			this.items = items;
-			this.type = type;
-		}
-
-		@Override
-		public T next() {
-			@SuppressWarnings("unchecked") // check is performed in hasNext
-			T result = (T) items[index + 1];
-			index += 2;
-			return result;
-		}
-
-		@Override
-		public boolean hasNext() {
-			if (items == null) {
-				return false;
-			}
-			while (index < items.length && items[index] != type) {
-				index += 2;
-			}
-			return index < items.length;
-		}
-
-		@Override
-		public void remove() {
-			throw new UnsupportedOperationException("Iterator removal not supported"); //$NON-NLS-1$
-		}
+		Deque<Object> specListeners = listeners.get(c);
+		return specListeners != null && !specListeners.isEmpty();
```

Comment:
Now that the Deque object is removed together with the last listener, I suppose you can go back to this simpler check, right?
```suggestion
		return listeners.containsKey(c);
```

Reply:
good point.